src/api_routes_websocket.rs

```diff
@@ -117,131 +131,207 @@ use std::{
   ops::Deref,
   result,
   str::FromStr,
-  sync::{Arc, Mutex},
   time::{Duration, Instant},
 };
-use tracing::{debug, error, info};
+use tracing::{debug, error};
 
-/// Entry point for our route
-pub async fn websocket(
-  req: HttpRequest,
-  body: web::Payload,
-  context: web::Data<LemmyContext>,
-  rate_limiter: web::Data<RateLimitCell>,
+/// How often heartbeat pings are sent
+const HEARTBEAT_INTERVAL: Duration = Duration::from_secs(5);
+
+/// How long before lack of client response causes a timeout
+const CLIENT_TIMEOUT: Duration = Duration::from_secs(10);
+
+pub struct WsChatSession {
+  /// unique session id
+  pub id: ConnectionId,
+
+  pub ip: IpAddr,
+
+  /// Client must send ping at least once per 10 seconds (CLIENT_TIMEOUT),
+  /// otherwise we drop connection.
+  pub hb: Instant,
+
+  /// Chat server
+  // pub cs_addr: Addr<ChatServer>,
+
+  /// A rate limiter for websocket joins
+  // rate_limiter: RateLimitCell,
+
+  // TODO this seems dangerous, but I can't find another way around it. parse_json_message
+  // needs the apub context
+  /// The context data
   apub_data: ContextData<LemmyContext>,
-) -> Result<HttpResponse, Error> {
-  let (response, session, stream) = actix_ws::handle(&req, body)?;
+}
```

Comment:
This is the main thing I need help with. I'd prefer to only have the Chat session require the `Addr<ChatServer>`, but `parse_json_message` still requires this apub context object.

Reply:
I dont see any problem here, certainly nothing dangerous. Just make sure you clone the apub_data every time you pass it into a request handler (should be unavoidable anyway), otherwise the counter for outgoing http requests will keep incrementing and federation will start failing soon.